### PR TITLE
UCP/API: add ucp_ep_modify_nb

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
@alinask @yosefe

### CODE REVIEW

**Comment by evgeny-leksikov** [comment]
This PR has a conflict with https://github.com/openucx/ucx/pull/2126

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5678/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5691/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
timed-out
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5694/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
https://github.com/openucx/ucx/issues/2118
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5700/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5703/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -807,13 +807,15 @@ typedef struct ucp_ep_params {
      * is not set, this address is mandatory for filling
      * (along with its corresponding bit in the field_mask - @ref
      * UCP_EP_PARAM_FIELD_REMOTE_ADDRESS) and must be obtained using
-     * @ref ucp_worker_get_address.
+     * @ref ucp_worker_get_address. This is not reconfigurable parameter by
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: `This field cannot be changed by @ref ucp_ep_modify_nb` (need to fix in all places)

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -1552,6 +1555,32 @@ ucs_status_t ucp_ep_create(ucp_worker_h worker, const ucp_ep_params_t *params,
                            ucp_ep_h *ep_p);
 
 
+/**
+ * @ingroup UCP_ENDPOINT
+ * @brief Modify endpoint parameters.
+ *
+ * This routine modifies @ref ucp_ep_h "endpoint" created by @ref ucp_ep_create
+ * or @ref ucp_listener_accept_callback_t.
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: Need to say that only some of the parameters can be modified and that the documentation of ucp_ep_params_t  species which exactly

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -1552,6 +1555,32 @@ ucs_status_t ucp_ep_create(ucp_worker_h worker, const ucp_ep_params_t *params,
...
+ *
+ * @param [in]  ep          A handle to the endpoint.
+ * @param [in]  params      User defined @ref ucp_ep_params_t configurations
+ *                          for the @ref ucp_ep_h "UCP endpoint".
+ *
+ * @return UCS_OK           - The endpoint is modified successfully.
+ * @return UCS_PTR_IS_ERR(_ptr) - The reconfiguration failed and an error code
+ *                                indicates the status. However, resources are
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: i think we don't want to kill the endpoint, but just keep the original configuration, right?

File: src/ucp/api/ucp.h [hunk 4]
```diff
@@ -1552,6 +1555,32 @@ ucs_status_t ucp_ep_create(ucp_worker_h worker, const ucp_ep_params_t *params,
...
+ * @brief Modify endpoint parameters.
+ *
+ * This routine modifies @ref ucp_ep_h "endpoint" created by @ref ucp_ep_create
+ * or @ref ucp_listener_accept_callback_t.
+ *
+ * @param [in]  ep          A handle to the endpoint.
+ * @param [in]  params      User defined @ref ucp_ep_params_t configurations
+ *                          for the @ref ucp_ep_h "UCP endpoint".
+ *
+ * @return UCS_OK           - The endpoint is modified successfully.
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: NULL instead of UCS_OK

File: src/ucp/api/ucp.h (hunk 3 above)
Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h]: I'm not sure :) this is possible only when reconfiguration was requested for local parameters. Yes, currently, we say that "remote" parameters are not reconfigurable and the only error that can be returned here is `UCS_ERR_INVALID_PARAM` and EP will not be changed. Maybe add this is as an exception? I also thought to add a note that if the function returned "request", the endpoint cannot be used until its completion. @yosefe  What do you think?

Comment (yosefe) [reply src/ucp/api/ucp.h]: i think the endpoint can be used but requests will be pending, as with connection establishment
also if there is failure the config remote params the endpoint can be moved to error state

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5748/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h (hunk 2 above)
Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h]: added a `@note` below

**Comment by evgeny-leksikov** [comment]
@alinask pls review

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5802/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@evgeny-leksikov pls answer https://github.com/openucx/ucx/pull/2130#discussion_r159457533

File: src/ucp/api/ucp.h (hunk 3 above)
Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h]: makes sense, i'll fix the description like "in case if error is returned, the endpoint is not changed."

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5805/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@shamisp @bbenton pls take a look

File: src/ucp/api/ucp.h [hunk 5]
```diff
@@ -844,7 +846,7 @@ typedef struct ucp_ep_params {
      * if the @ref UCP_EP_PARAMS_FLAGS_CLIENT_SERVER flag is set, this address
      * is mandatory for filling (along with its corresponding bit in the
      * field_mask - @ref UCP_EP_PARAM_FIELD_SOCK_ADDR) and should be obtained
-     * from the user.
+     * from the user. This field cannot be changed by @ref ucp_ep_modify_nb.
      */
     ucs_sock_addr_t         sockaddr;
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: do we have the check and error in ucp_ep_modify_nb ?

File: src/ucp/api/ucp.h (hunk 2 above)
Comment (shamisp) [reply src/ucp/api/ucp.h]: Describe the use case for the function

File: test/gtest/ucp/test_ucp_peer_failure.cc [hunk 6]
```diff
@@ -134,6 +145,21 @@ void test_ucp_peer_failure::init() {
         smoke_test();
     }
     wrap_errors();
+
+    ucp_ep_params_t ep_params_mod = {0};
+    ep_params_mod.field_mask = UCP_EP_PARAM_FIELD_ERR_HANDLER_CB |
+                               UCP_EP_PARAM_FIELD_USER_DATA;
```
Comment (shamisp) [inline test/gtest/ucp/test_ucp_peer_failure.cc]: Where do we test the error flow ? (attempt to update a field that it is cost)?

File: src/ucp/api/ucp.h (hunk 5 above)
Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h]: Yes, see https://github.com/openucx/ucx/pull/2130/files#diff-8fd5f37669b4533d1e9389ff6839dcccR400

File: test/gtest/ucp/test_ucp_peer_failure.cc (hunk 6 above)
Comment (evgeny-leksikov) [reply test/gtest/ucp/test_ucp_peer_failure.cc]: Do you mean this (https://github.com/openucx/ucx/pull/2130/files/7ac6bc63bf7d2e8a9775ef8578ed6833e3d783af#diff-4fb58bb922ffb07d1d1682f0129c0aa4R179)?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5824/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
@bbenton please take a look

**Comment by bbenton** [comment]
:+1: for doc/API

**Pull request merged**